kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraWindowedTable.java

```
@@ -428,47 +488,112 @@ public String name() {
+  //  instead, or whether TaskMigratedException thrown here will be properly handled by Streams
+  private void handleEpochFencing(
+      final int kafkaPartition,
+      final SegmentPartition tablePartition,
+      final long epoch
+  ) {
+    final long otherEpoch = fetchEpoch(tablePartition);
+    final var msg = String.format(
+        "Could not initialize commit buffer [%s-%d] - attempted to claim epoch %d, "
+            + "but was fenced by a writer that claimed epoch %d on table partition %s",
+        name(),
+        kafkaPartition,
+        epoch,
+        otherEpoch,
+        tablePartition
+    );
+    final var e = new TaskMigratedException(msg);
+    LOG.warn(msg, e);
+    throw e;
+  }
+
   @Override
   public void preCommit(
       final int kafkaPartition,
       final long epoch
   ) {
-    if (pendingFlushStreamTime > lastFlushStreamTime) {
-      activeRoll = partitioner.rolledSegments(name, lastFlushStreamTime, pendingFlushStreamTime);
+    final PendingFlushInfo pendingFlush = kafkaPartitionToPendingFlushInfo.get(kafkaPartition);
+    final SegmentRoll pendingRoll = partitioner.rolledSegments(
+        name, pendingFlush.lastFlushStreamTime, pendingFlush.pendingFlushStreamTime
+    );
 
-      for (final long segmentId : activeRoll.segmentsToCreate) {
-        initSegment(new SegmentPartition(kafkaPartition, segmentId), epoch);
+    pendingFlush.initSegmentRoll(pendingRoll);
+    for (final long segmentId : pendingRoll.segmentsToCreate) {
+      final SegmentPartition segment = new SegmentPartition(kafkaPartition, segmentId);
+      final var createSegment = createSegment(segment, epoch);
+
+      // If the segment creation failed because the table partition already exists, attempt to
+      // update the epoch in case we are fencing an older writer -- if that fails it means we're
+      // the ones being fenced
+      // TODO: what if the segment creation failed for a reason besides already existing?
```

then it will throw an error instead of returning `!wasApplied`

---

oh, that's good to know 😄 I'll remove the TODO